### PR TITLE
hardening tool: Update recommendations

### LINKED ISSUE #57166: hardening: hardened.csv is outdated
**Describe the bug**

Zephyr's hardening tool might be one of those hidden gems that many folks (including maintainers) don't even know exist.
https://docs.zephyrproject.org/latest/security/hardening-tool.html#usage

The [hardened.csv](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/kconfig/hardened.csv) file is meant to help with flagging the use of features not recommended for production devices (ex. experimental features, logging, ...).
It appears that the file hasn't been maintained and several features that moved out of experimentals (NET_SOCKETS_OFFLOAD, UART_ASYNC_API, ...) are still marked experimental there. What's more, there's probably many other features added over the years that'd need to be properly documented w.r.t their recommended values.

**To Reproduce**

1.  `west build -b nrf52840dk_nrf52840 .\samples\subsys\logging\logger` (a samples that uses UART_ASYNC_API)
2. `cd build`
3. `ninja hardenconfig`
4. FAIL due to CONFIG_UART_ASYNC_API being set to Y when recommend is N. But this API is not experimental anymore so shouldn't be "not" recommended.

**Expected behavior**

`hardened.csv` file should reflect actual recommended values as per maintainers/security group, in particular all the experimental/beta features should be aligned with current codebase.

**Impact**
Lack of clarity w.r.t relevance/usefulness of the hardening tool.

**Environment (please complete the following information):**
 - OS: all
 - Toolchain: 0.16.0
 - Commit SHA or Version: 3a08d9620ae5528d36f3b013b906eadfc9fed5e7

**Additional context**

cc @ceolin @kestewart

### PR DESCRIPTION
**Pull request by ceolin**
Update general recommendations in the hardening tool.

Fixes: #57166

### CODE REVIEW

File: scripts/kconfig/hardened.csv [hunk 1]
```diff
@@ -1,85 +1,147 @@
...
+MODEM_CONTEXT,n,experimental
+MQTT_SN_LIB,n,experimental
 NET_BUF_VARIABLE_DATA_SIZE,n,experimental
 NET_CONNECTION_MANAGER,n,experimental
 NET_GPTP,n,experimental
+NET_GPTP,n,experimental
 NET_IPV4_AUTO,n,experimental
 NET_L2_IEEE802154_SECURITY,n,experimental
+NET_L2_IEEE802154_SECURITY,n,experimental
+NET_L2_IEEE802154_SECURITY,n,experimental
```
Comment (kartben) [inline scripts/kconfig/hardened.csv:106]: entry was already there?

File: scripts/kconfig/hardened.csv [hunk 2]
```diff
@@ -1,85 +1,147 @@
...
 NET_L2_PPP,n,experimental
-NET_L2_PTP,n,experimental
-NET_OFFLOAD,n,experimental
 NET_PROMISCUOUS_MODE,n,experimental
 NET_SOCKETS_CAN,n,experimental
 NET_SOCKETS_ENABLE_DTLS,n,experimental
+NET_SOCKETS_ENABLE_DTLS,n,experimental
```
Comment (kartben) [inline scripts/kconfig/hardened.csv:111]: entry was already there?

File: scripts/kconfig/hardened.csv [hunk 3]
```diff
@@ -1,85 +1,147 @@
...
+LOG_BACKEND_BLE,n,experimental
+LORA,n,experimental
+LORAWAN,n,experimental
+LWM2M_FIRMWARE_UPDATE_OBJ_SUPPORT_MULTIPLE,n,experimental
+LWM2M_GATEWAY_OBJ_SUPPORT,n,experimental
+MEMC,n,experimental
+MIPI_DSI,n,experimental
+MM_DRV,n,experimental
 MODEM_CONTEXT,n,experimental
+MODEM_CONTEXT,n,experimental
```
Comment (kartben) [inline scripts/kconfig/hardened.csv:97]: entry was already there?

File: scripts/kconfig/hardened.csv [hunk 4]
```diff
@@ -1,85 +1,147 @@
...
+MQTT_SN_LIB,n,experimental
 NET_BUF_VARIABLE_DATA_SIZE,n,experimental
 NET_CONNECTION_MANAGER,n,experimental
 NET_GPTP,n,experimental
+NET_GPTP,n,experimental
```
Comment (kartben) [inline scripts/kconfig/hardened.csv:102]: entry was already there?

File: scripts/kconfig/hardened.csv [hunk 5]
```diff
@@ -1,85 +1,147 @@
...
+NET_SOCKETS_ENABLE_DTLS,n,experimental
 NET_SOCKETS_NET_MGMT,n,experimental
-NET_SOCKETS_OFFLOAD,n,experimental
 NET_SOCKETS_SOCKOPT_TLS,n,experimental
-DEBUG_THREAD_INFO,n,experimental
-PERFORMANCE_METRICS,n,experimental
+NET_TC_THREAD_PREEMPTIVE,n,experimental
+OPENTHREAD_MLE_LONG_ROUTES,n,experimental
 SHELL_TELNET_SUPPORT_COMMAND,n,experimental
+SHELL_TELNET_SUPPORT_COMMAND,n,experimental
```
Comment (kartben) [inline scripts/kconfig/hardened.csv:117]: entry was already there?

**Review by kartben — CHANGES_REQUESTED** [review]
Very nice, thanks for doing this! 
I just had a quick look for now as I spotted several double/triple entries so I'm thinking that should you have used some kind of automated tool, you might need to double check things before this goes through a more thorough review.

File: scripts/kconfig/hardened.csv [hunk 6]
```diff
@@ -1,6 +1,5 @@
 HW_STACK_PROTECTION,y
```
Comment (stephanosio) [inline scripts/kconfig/hardened.csv]: would be nice to add a comment at the top telling whoever modifies this file in the future to keep the list sorted alphabetically.

Comment (kartben) [reply scripts/kconfig/hardened.csv]: +1, but this is making me realize that the file supposedly being "CSV", I am not so sure whether we want comments in the first place. I know that I would never thought of adding a comment such as https://github.com/zephyrproject-rtos/zephyr/blob/bfdfbd92d6854da8273826ace1fa8ca9ae81eb68/scripts/kconfig/hardened.csv#L49 myself, for the very reason that I would not expect it to work. It turns out that this probably ends up caught [here](https://github.com/zephyrproject-rtos/zephyr/blob/main/scripts/kconfig/hardenconfig.py#L51-L54) but this feels hackish, no?
But don't get me wrong, I definitely like the conciseness of a CSV-like syntax in this particular instance.

Comment (ceolin) [reply scripts/kconfig/hardened.csv]: Regarding comments, it is catch in that exception, it is not different from an empty line. I can add some logic to explicitly handle it but it seems an unnecessary overhead.

I put these comments to separate some logical sections, I think it makes it easier to review it.

File: scripts/kconfig/hardened.csv (hunk 3 above)
Comment (ceolin) [reply scripts/kconfig/hardened.csv]: fixed

**Review by tejlmand — APPROVED** [review]

**Comment by ceolin** [comment]
@kartben  @stephanosio can you take a look again on it ? If you are not satisfied with the answer about comments in that file I can remove them.

**Comment by kartben** [comment]
FWIW I think it would make sense to also update the documentation page as it could use a refresh (ex. usage instructions). https://docs.zephyrproject.org/latest/security/hardening-tool.html#hardening-tool

It might also be worthwhile to indicate that an option to run the hardenconfig tool is simply to run `west build -t hardenconfig`

**Review by kartben — APPROVED** [review]

**Review by stephanosio — APPROVED** [review]

**Comment by kartben** [comment]
> FWIW I think it would make sense to also update the documentation page as it could use a refresh (ex. usage instructions). https://docs.zephyrproject.org/latest/security/hardening-tool.html#hardening-tool
> 
> It might also be worthwhile to indicate that an option to run the hardenconfig tool is simply to run `west build -t hardenconfig`

Addressed this in https://github.com/zephyrproject-rtos/zephyr/pull/57958

**Pull request merged**